### PR TITLE
Add support for CURLOPT_RESOLVER_START callback

### PR DESCRIPTION
**Pull request by fsedano**
This patch adds a callback which will be called when libcurl tries to start a new resolve, passing the resolver data.

If using ARES, this allows the user to access the ARES channel and configure callbacks or other options.

### CODE REVIEW

**Comment by bagder** [comment]
The files `CURLOPT_RESOLVER_START_DATA.3` and `CURLOPT_RESOLVER_START_FUNCTION.3` are missing from the dist tarball - which causes the CI distcheck test to fail. Add them in docs/libcurl/opts/Makfile.inc.

File: docs/libcurl/opts/CURLOPT_RESOLVER_START_FUNCTION.3 [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+.\" **************************************************************************
+.\"
+.TH CURLOPT_RESOLVER_START_FUNCTION 3 "14 Feb 2018" "libcurl 7.59.1" "curl_easy_setopt options"
+.SH NAME
+CURLOPT_RESOLVER_START_FUNCTION \- set callback to be called before resolver starts
+.SH SYNOPSIS
+.nf
+#include <curl/curl.h>
+
+void resolver_alloc_cb (void *resolver_data, void *user_data);
```
Comment (bagder) [inline docs/libcurl/opts/CURLOPT_RESOLVER_START_FUNCTION.3]: I think you need to allow this function to return *at least* an error for whatever went wrong in the callback that makes the code want to stop the transfer. So maybe just an int and let 0 mean OK, 1 is means "error, stop, get out" and the rest left reserved for the future?

File: docs/libcurl/opts/CURLOPT_RESOLVER_START_FUNCTION.3 [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+No callback will be made
+.SH PROTOCOLS
+For all protocols
+.SH AVAILABILITY
+Support for CURLOPT_RESOLVER_START_FUNCTION was added in version 7.59.1.
+.SH RETURN VALUE
+This will return CURLE_OK.
+.SH EXAMPLE
+A common use case is to use this callback to setup any options on the ares
+channel, for example socket callbacks.
```
Comment (bagder) [inline docs/libcurl/opts/CURLOPT_RESOLVER_START_FUNCTION.3]: I think the **EXAMPLE** section should show actual code using the option.

File: docs/libcurl/opts/CURLOPT_RESOLVER_START_DATA.3 [hunk 3]
```diff
@@ -0,0 +1,48 @@
+.\" **************************************************************************
+.\" *                                  _   _ ____  _
+.\" *  Project                     ___| | | |  _ \| |
+.\" *                             / __| | | | |_) | |
+.\" *                            | (__| |_| |  _ <| |___
+.\" *                             \___|\___/|_| \_\_____|
+.\" *
+.\" * Copyright (C) 1998 - 2015, Daniel Stenberg, <daniel@haxx.se>, et al.
```
Comment (bagder) [inline docs/libcurl/opts/CURLOPT_RESOLVER_START_DATA.3]: I think using 2018 as the end year in the copyright year range from start seems appropriate! =)

File: docs/libcurl/opts/CURLOPT_RESOLVER_START_FUNCTION.3 [hunk 4]
```diff
@@ -0,0 +1,60 @@
+.\" **************************************************************************
+.\" *                                  _   _ ____  _
+.\" *  Project                     ___| | | |  _ \| |
+.\" *                             / __| | | | |_) | |
+.\" *                            | (__| |_| |  _ <| |___
+.\" *                             \___|\___/|_| \_\_____|
+.\" *
+.\" * Copyright (C) 1998 - 2014, Daniel Stenberg, <daniel@haxx.se>, et al.
```
Comment (bagder) [inline docs/libcurl/opts/CURLOPT_RESOLVER_START_FUNCTION.3]: end year

File: include/curl/curl.h [hunk 5]
```diff
@@ -1823,6 +1827,13 @@ typedef enum {
      seconds since 1 Jan 1970. */
   CINIT(TIMEVALUE_LARGE, OFF_T, 270),
 
+  /* Function that will be called before a resolver request is made */
+  CINIT(RESOLVER_START_FUNCTION, FUNCTIONPOINT, 271),
+
+  /* User data to pass to the resolver start callback */
+  CINIT(RESOLVER_START_DATA, OBJECTPOINT, 272),
+
+
```
Comment (bagder) [inline include/curl/curl.h]: extra newline, let's stick to only one between each option

File: lib/hostasyn.c [hunk 6]
```diff
@@ -147,6 +147,15 @@ Curl_addrinfo *Curl_getaddrinfo(struct connectdata *conn,
                                 int port,
                                 int *waitp)
 {
+  struct Curl_easy *data = conn->data;
+
+  /* Call the resolver start callback, if defined */
+  if(data->set.resolver_callbacks.resolver_start_cb) {
+    data->set.resolver_callbacks.
```
Comment (bagder) [inline lib/hostasyn.c]: so here I would argue there should be a return code and handling of that, as mentioned above

**Comment by bagder** [comment]
After I merged b46cfbc06 just now, I would also like to have the callback work like this:

      Curl_set_in_callback(conn->data, true);
      rc = invoke_callback ( args ...) ;
      Curl_set_in_callback(conn->data, false);

to make us detect attempts of recursive libcurl calls.

File: lib/hostasyn.c (hunk 6 above)
Comment (fsedano) [reply lib/hostasyn.c]: Sure, thanks for the suggestion - I've fixed the failing TCs and added the example, will add the two other issues you mention (recursive CB attempt + return code) on a subsequent commit. Thanks for the help.

**Comment by fsedano** [comment]
Is the breakage with travis known? It does not look related to the patch.

**Comment by bagder** [comment]
The travis jobs sometimes time out/fail in ways that aren't our fault. I restarted that single failed job now.

File: lib/hostasyn.c (hunk 6 above)
Comment (fsedano) [reply lib/hostasyn.c]: Just submitted code with the requested changes, updated TCs and doc.
Thanks,

**Comment by fsedano** [comment]
Seems to fail due to a parent issue:
 c:\projects\curl\lib\hostip.c(957): warning C4701: potentially uninitialized local variable 'addresses' used [C:\projects\curl\build.msvc2012\lib\libcurl.vcxproj]
  c:\projects\curl\lib\hostip.c(957): warning C4703: potentially uninitialized local pointer variable 'addresses' used [C:\projects\curl\build.msvc2012\lib\libcurl.vcxproj]

**Comment by fsedano** [comment]
It seems travis failed again for something unrelated to us. Can we have this restarted please? We need this patch to be part of next release

**Comment by bagder** [comment]
"This branch has conflicts that must be resolved" due to conflicting changes that were merged recently. The unrelated errors were also fixed in the mean time I believe.

**Comment by fsedano** [comment]
Working on the conflicts

**Comment by fsedano** [comment]
Everything looks fine - Could we have this merged for 7.59.0 please? It's important for us.

**Comment by bagder** [comment]
I'm really sorry, but I rather err on the safe side rather than hurry to get something in before the curtain falls. I have not had time yet to check this out closer and properly consider the API and function arguments etc - and nobody else has stepped up with feedback (neither positive nor negative). I need more time, so it will not get merged in this feature window. I apologize for this, but I've had a few very busy weeks at work lately and it has made me had less time to curl.

**Comment by jay** [comment]
@bagder I started reviewing it this afternoon. I'm changing some of the doc and internal naming to bring it in line. I can have it in before the deadline, let me know if that is a problem.

**Comment by bagder** [comment]
@jay, I fully trust your judgement so if you're happy with it and think its fine to merge, I'm just happy. Go ahead!

**Comment by fsedano** [comment]
@Jay, @bagder really helpful. As mentioned, this feature is crucial for our project and we need to use an official curl version, so it's important for us. Please let me know if I can be of any help.

Thanks,

**Comment by jay** [comment]
I moved the callback into Curl_resolv right before Curl_getaddrinfo is called. This way it is called for all backends, even though currently it appears only to have a benefit for the ares backend. I also added a 'reserved' pointer to the callback in case we want to expand on it in the future. (For example allowing the user to receive the hostname/port and do their own resolve maybe?)

~~~c
int resolver_start_cb(void *resolver_state, void *reserved, void *userdata);
~~~

The EXAMPLE in the doc I don't think is sufficient since it just shows resolver_state pointer address. It could really use an example that shows how the resolver_state should be used with ares, like if multiple hostnames are resolved do you set the socket multiple times? etc, or how in practice it is used.

**Pull request closed without merging**